specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,143 @@
+be limited to prevent a denial of service attack against L2 as well as allow the
+total amount of guaranteed gas to be below the L2 block gas limit.
+
+We set limit the total amount of gas buyable via a contract method. It will initially
+be controlled by the Optimism Multisig before being handed over to governance.
+TODO - check that this is the actual plan.
+
+## 1559 Fee Market
+
+To reduce [Priority Gas Auctions](./glossary.md#priority-gas-auction) and accurately price gas, we implement a 1559
+style fee market on L1 with the following pseudocode. We also use this opporunity to
+place a hard limit on the amount of guaranteed gas that is provided.
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+GUARANTEED_GAS_LIMIT: u64, SANITY_GAS_LIMIT: u64 = load_and_unpack_storage2()
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
+    if curr_bought_gas == gas_target {
+        new_basefee := curr_basefee
+    } else if curr_bought_gas > gas_target {
+        gas_delta     := curr_bought_gas - gas_target
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        basefee_delta := max(basefee_delta, 1) # TODO: Why does 1559 have this asymmetry?
+        new_basefee   := curr_basefee + basefee_delta
+    } else {
+        gas_delta     := gas_target - curr_bought_gas
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+    # Fun fact, geth doesn't let the new_basefee get below 0 and while not in the EIP spec, we should add this as well.
+        new_basefee   := curr_basefee - basefee_delta
+    }
+    curr_basefee := new_basefee
+    curr_number := block.number
+    curr_bought_gas := 0
+   
+}
+
+curr_bought_gas += required_gas
+require(curr_bought_gas <= min(GUARANTEED_GAS_LIMIT, SANITY_GAS_LIMIT)
+gas_cost = requested_gas * curr_basefee
+
+burn(gas_cost) OR pay_to_contract(gas_cost) # Depends if payable or non-payable version
+
+pack_and_store(curr_basefee, curr_number, curr_bought_gas)
+```
+
+```python
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+
+prev_basefee, prev_num, prev_bought_gas = load_and_unpack_storage()
+
+gas_target_limit, gas_sanity_limit = load_and_unpack_storage2()
+
+
+gas_cost = requested_gas * prev_basefee
+
+now_basefee = prev_basefee
+now_num = block.number
+now_bought_gas = prev_bought_gas + requested_gas
+# update only if we are in a new block
+if now_num != prev_num:
+  now_bought_gas = requested_gas
+  # update basefee if we are using more or less than the target
+  if prev_bought_gas < gas_target:
+      gas_used_delta = prev_bought_gas - gas_target
```

Comment:
There is an underflow here here, if `prev_bought_gas` is smaller then it must be on the right hand side of the subtraction operator